Update store setup widget to use task list API

joshuatf opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Update the widget visibility capabilities and use the new task list API to retrieve tasks instead of hard-coding them.

### How to test the changes in this Pull Request:

#### Testing capabilities

1. Start with a fresh site with an unfinished task list
2. Go to the WordPress Dashboard page
3. Note as an admin that you can see the `WooCommerce Setup` widget
4. Create a user role that should not be able to manage WooCommerce but can view the dashboard (e.g., Author)
5. Log in with that user and navigate to the dashboard.
6. Note that the `WooCommerce Setup` widget is not visible

#### Testing tasks

1. Navigate to the WordPress dashboard
2. Note that the correct number of tasks are shown (complete/total) for the `WooCommerce Setup` widget
3. Click on the action button `Start selling` and make sure you’re redirected to the correct page
4. Repeat this process for multiple tasks, taking special attention for the “store details/setup” task since it uses a custom URL passed via the API

### Changelog entry

> Update store setup widget to use task list API

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshuatf committed:
Hide store setup widget if user cannot manage woocommerce

joshuatf committed:
Use task list API to retrieve tasks

joshuatf committed:
Fix up URL and URL fragments from tasks

octaedro reviewed (dismissed):
Good job @joshuatf! This is testing well on my end and code LGTM 🚀

joshuatf committed:
Remove URL assertion on external API response

jessepearson reviewed (changes requested):
Everything works well, as an Author I can no longer see the widget when logged in. 👍 

The only thing I noticed is that the numbers in the widget do not match up with the steps. For instance, on a fresh site it says "Step 0 of 7". After completing the first step it has "1 of 7":

![Screen Shot 2022-01-28 at 3 07 55 PM](https://user-images.githubusercontent.com/4634416/151606994-72537564-fd0e-4d60-92be-b29e01b17554.png)

The logic here may be that "Step 1 of 7" is complete, like shown in the next screenshot, but from my experience with UIs I believe they typically tell you which number you are on.
![Screen Shot 2022-01-28 at 3 07 39 PM](https://user-images.githubusercontent.com/4634416/151607167-de2c352a-354a-48d6-8537-c4e585ee8a10.png)

joshuatf committed:
Simplify task count and completed count test

joshuatf committed:
Update tests around widget visibility

joshuatf committed:
Update step number to be completed count +1

joshuatf commented:
Thanks, @jessepearson - agree with your comments and updated to reflect the step of the item we're on as opposed to completed items.

Also fixed up the tests and made some minor updates in logic to the widget class to allow for easier testing.  On the off chance that anyone is on for the weekend, this one needs another review before we merge.  cc @octaedro @woocommerce/isotope

octaedro reviewed (approved):
LGTM! :shipit:

wplaunchify commented:
This update throws a critical error in my multisite setup. See below:
When I go back to previous commit the error is resolved.

Fatal error: Uncaught Error: Call to a member function is_complete() on null in /home/387089.cloudwaysapps.com/jwmkkrtvnr/public_html/wp-content/plugins/woocommerce/includes/admin/class-wc-admin-dashboard-setup.php:173 Stack trace: #0 /home/387089.cloudwaysapps.com/jwmkkrtvnr/public_html/wp-content/plugins/woocommerce/includes/admin/class-wc-admin-dashboard-setup.php(44): WC_Admin_Dashboard_Setup->should_display_widget() #1 /home/387089.cloudwaysapps.com/jwmkkrtvnr/public_html/wp-content/plugins/woocommerce/includes/admin/class-wc-admin-dashboard-setup.php(181): WC_Admin_Dashboard_Setup->__construct() #2 /home/387089.cloudwaysapps.com/jwmkkrtvnr/public_html/wp-content/plugins/woocommerce/includes/admin/class-wc-admin.php(102): include('/home/387089.cl...') #3 /home/387089.cloudwaysapps.com/jwmkkrtvnr/public_html/wp-includes/class-wp-hook.php(307): WC_Admin->conditional_includes() #4 /home/387089.cloudwaysapps.com/jwmkkrtvnr/public_html/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters() #5 /home/387089.cloudwaysa in /home/387089.cloudwaysapps.com/jwmkkrtvnr/public_html/wp-content/plugins/woocommerce/includes/admin/class-wc-admin-dashboard-setup.php on line 173
There has been a critical error on this website.

joshuatf commented:
Hi @wplaunchify - thanks for bringing this to our attention.  Would you mind opening an issue [here](https://github.com/woocommerce/woocommerce/issues/new/choose) so that we can address this as soon as possible?

If there are any other steps needed to reproduce, please list them in the issue.
